pkg/events/events.go

```diff
@@ -0,0 +1,275 @@
+}
+
+type Subscription interface {
+	Unsubscribe()
+	Err() <-chan error
+}
+
+type Listener struct {
+	logger        *slog.Logger
+	evmClient     EVMClient
+	progressStore ProgressStore
+	subMu         sync.RWMutex
+	subscribers   map[common.Hash][]*subscription
+	contracts     map[common.Address]*abi.ABI
+}
+
+func NewListener(
+	logger *slog.Logger,
+	evmClient EVMClient,
+	progressStore ProgressStore,
+	contracts map[common.Address]*abi.ABI,
+) *Listener {
+	return &Listener{
+		logger:        logger,
+		evmClient:     evmClient,
+		progressStore: progressStore,
+		subscribers:   make(map[common.Hash][]*subscription),
+		contracts:     contracts,
+	}
+}
+
+type subscription struct {
+	event EventHandler
+	unsub func()
+	errCh chan error
+}
+
+func (s *subscription) Unsubscribe() {
+	s.unsub()
+}
+
+func (s *subscription) Err() <-chan error {
+	return s.errCh
+}
+
+func (l *Listener) Subscribe(event EventHandler) (Subscription, error) {
+	found := false
+	for _, c := range l.contracts {
+		for _, e := range c.Events {
+			if e.Name == event.EventName() {
+				event.SetTopicAndContract(e.ID, c)
+				found = true
+				break
+			}
+		}
+	}
+
+	if !found {
+		return nil, fmt.Errorf("event not found")
+	}
+
+	l.subMu.Lock()
+	defer l.subMu.Unlock()
+
+	sub := &subscription{
+		event: event,
+		errCh: make(chan error),
+		unsub: func() { l.unsubscribe(event) },
+	}
+
+	l.subscribers[event.Topic()] = append(l.subscribers[event.Topic()], sub)
+
+	return sub, nil
+}
+
+func (l *Listener) unsubscribe(event EventHandler) {
+	l.subMu.Lock()
+	defer l.subMu.Unlock()
+
+	events := l.subscribers[event.Topic()]
+	for i, e := range events {
+		if e.event == event {
+			events = append(events[:i], events[i+1:]...)
+			break
+		}
+	}
+
+	l.subscribers[event.Topic()] = events
+}
+
+func (l *Listener) publishLogEvent(ctx context.Context, log types.Log) {
+	l.subMu.RLock()
+	defer l.subMu.RUnlock()
+
+	events := l.subscribers[log.Topics[0]]
+	for _, event := range events {
+		ev := event
+		go func() {
+			if err := ev.event.Handle(log); err != nil {
+				l.logger.Error("failed to handle log", "error", err)
+				select {
+				case ev.errCh <- err:
+				case <-ctx.Done():
+				}
+			}
+		}()
+	}
+}
+
+func (l *Listener) Start(ctx context.Context) <-chan struct{} {
+	doneChan := make(chan struct{})
+
+	if len(l.contracts) == 0 {
+		close(doneChan)
+		return doneChan
+	}
+
+	go func() {
+		defer close(doneChan)
+
+		lastBlock, err := l.progressStore.LastBlock()
+		if err != nil {
+			l.logger.Error("failed to get last block", "error", err)
+			return
+		}
+
+		contracts := make([]common.Address, 0, len(l.contracts))
+		for addr := range l.contracts {
+			contracts = append(contracts, addr)
+		}
+
+		q := ethereum.FilterQuery{
+			FromBlock: big.NewInt(int64(lastBlock + 1)),
+			ToBlock:   nil,
+			Addresses: contracts,
+		}
+
+		logChan := make(chan types.Log)
+
+		sub, err := l.evmClient.SubscribeFilterLogs(ctx, q, logChan)
```

Comment:
The EvmClient needs to be initialized with a WebSocket endpoint; otherwise, it will return a 'subscription is not supported' error. I faced this issue before